### PR TITLE
Support arbitrarily defined revision ranges

### PR DESCRIPTION
**Pull request by joba00002**
Addresses #36

### Prerequisites

- [x] I adhere to the coding conventions (described [here](../CONTRIBUTING.md)) in my code.
- [x] I have updated the copyright headers of the files I have modified.
- [x] I have written appropriate commit messages, i.e., I have recorded the goal, the need, the needed changes, and the location of my code modifications for each commit. This includes also, e.g., referencing to relevant issues.
- [x] I have put signed-off tags in *all* commits.
- [x] I have updated the changelog file [NEWS.md](../NEWS.md) appropriately.
- [x] I have checked whether I need to adjust the showcase file `showcase.R` with respect to my changes.
- [x] The pull request is opened against the branch `dev`.

### CODE REVIEW

File: util-data.R [hunk 1]
```diff
@@ -173,6 +176,8 @@ ProjectData = R6::R6Class("ProjectData",
         pasta.commits = create.empty.pasta.list(), # data.frame
         ## timestamps of mail, issue and commit data
         data.timestamps = data.frame(start = numeric(0), end = numeric(0)), # data.frame
+        ## custom timestamps for splitting
+        custom.event.timestamps = list(),
```
Comment (bockthom) [inline util-data.R:180]: Please add the comment "# list" to be consistent with the other lines (even if this is obvious here that this is a list)

File: tests/test-data.R [hunk 2]
```diff
@@ -420,6 +421,33 @@ test_that("Filter bots from mail data", {
...
+    timestamps.events = proj.data$get.custom.event.timestamps()
+    timestamps.events.expected = list(
+        'Test event 1' = "2016-07-12 15:00:00",
+        'Test event 2' = "2016-07-12 16:00:00",
+        'Test event 3' = "2016-07-12 16:05:00",
+        'Test event 4' = "2016-10-05 09:00:00"
+    )
+
+    expect_identical(timestamps.events, timestamps.events.expected, "Custom timestamps.")
+})
```
Comment (bockthom) [inline tests/test-data.R:449]: Can we also test afterwards whether clearing the timestamps works correctly? So just updating the conf in such a way that the timestamps are cleared?

**Review by bockthom — CHANGES_REQUESTED** [review]
This PR looks good so far. I have had a look at all the changes (except for the splitting test, for which I need to have a bit more time). Please see my minor comments attached.

File: tests/README.md [hunk 3]
```diff
@@ -16,6 +16,7 @@ We have two test projects you can use when writing your tests:
      * Commit messages
      * Pasta
      * Synchronicity
+     * Custom event timestamps in `custom-events.list`
```
[the remark this answers is not in the document]
Comment (joba00002) [reply tests/README.md:19]: It does fail when "revisions.list" is empty, as an empty file also produces a `try-error`.

File: README.md [hunk 4]
```diff
@@ -589,6 +590,12 @@ There is no way to update the entries, except for the revision-based parameters.
     * The time-window (in days) to use for synchronicity data if enabled by `synchronicity = TRUE`
     * [1, *5*, 10, 15]
     * **Note**: If, at least, one artifact in a commit has been edited by more than one developer within the configured time window, then the whole commit is considered to be synchronous.
+- `custom.event.timestamps.file`:
+    * The file to read custom timestamps from.
+    * **Note** It might make sense to keep several lists of timestamps for different purposes. Therefore, this is the only data source where the file name can be configured.
```
Comment (bockthom) [inline README.md:595]: Maybe add another note that there is no default value for this parameter?

File: util-split.R [hunk 5]
```diff
@@ -55,9 +55,6 @@ requireNamespace("lubridate") # for date conversion
 #'                    [default: "commits"]
 #' @param sliding.window logical indicating whether the splitting should be performed using a sliding-window approach
 #'                       [default: FALSE]
-#' @param use.custom.events logical indicating whether custom event timestamps stored in \code{project.data} should be used.
-#'                       When used, all other parameters except \code{project.data} and \code{project.conf}.new are ignored.
-#'                       [default: FALSE]
```
Comment (bockthom) [inline util-split.R]: Could you please not only remove parameter documentation, but also the parameter itself from the function and its implementation?

File: util-split.R [hunk 6]
```diff
@@ -252,6 +249,31 @@ split.data.time.based = function(project.data, time.period = "3 months", bins =
...
 
+#' Split project data by timestamps
+#'
+#' Splits project data into ranges, where the first range starts with the first timestamp
+#' and the last range ends with the last timestamp.
+#'
+#' If timestamps are not provided, the custom timestamps in \code{project.data} are
+#' used instead.
+#'
+#' @param project.data the *Data object from which the data is retrieved
```
Comment (bockthom) [inline util-split.R:260]: Why does the "*" come from?

File: util-split.R [hunk 7]
```diff
@@ -252,6 +249,31 @@ split.data.time.based = function(project.data, time.period = "3 months", bins =
...
+#' @param bins a vector of timestamps
+#'             [default: NULL]
```
Comment (bockthom) [inline util-split.R:262]: Could you please move the default value to the end of the previous line?

**Review by bockthom — CHANGES_REQUESTED** [review]
Looks good, I just spotted a few minor issues; please see my comments below.

File: util-split.R [hunk 8]
```diff
@@ -252,6 +249,31 @@ split.data.time.based = function(project.data, time.period = "3 months", bins =
     return(cf.data)
 }
 
+#' Split project data by timestamps
+#'
+#' Splits project data into ranges, where the first range starts with the first timestamp
+#' and the last range ends with the last timestamp.
+#'
+#' If timestamps are not provided, the custom timestamps in \code{project.data} are
```
Comment (bockthom) [inline util-split.R:257]: custom timestamps --> custom event timestamps

File: tests/test-split-data-time-based.R [hunk 9]
```diff
@@ -987,8 +987,7 @@ patrick::with_parameters_test_that("Split a data object time-based ( use.custom.
     )
 
     ## split data
-    results = split.data.time.based(project.data, use.custom.events = TRUE,
-                                    split.basis = "mails")
+    results = split.data.time.based.by.timestamps(project.data)
```
Comment (bockthom) [inline tests/test-split-data-time-based.R:990]: Maybe adjust the comment to "split data by custom event timestamps" to make more obvious what is tested here.

File: util-split.R (hunk 6 above)
Comment (joba00002) [reply util-split.R:260]: I'm not sure either, but this is consistent with the other splitting functions.

File: util-data.R [hunk 10]
```diff
@@ -2087,6 +2101,44 @@ ProjectData = R6::R6Class("ProjectData",
...
             return (data)
+        },
+
+        #' Get the list of custom event timestamps,
+        #' read from a file configured by the \code{custom.event.timestamps.file}
+        #' parameter in the project configuration.
+        #'
+        #' @return the list of custom event timestamps
+        get.custom.event.timestamps = function() {
+            if (!self$is.data.source.cached("custom.event.timestamps") && !private$project.conf$get.value("custom.event.timestamps.locked")) {
```
Comment (hechtlC) [inline util-data.R:2112]: This line is too long as well.

**Review by hechtlC — CHANGES_REQUESTED** [review]
Looks good so far thanks @joba00002!
I just have some minor issues with line length that have to be addressed. Please do that and then we can talk tomorrow.

**Review by hechtlC — DISMISSED** [review]

File: NEWS.md [hunk 11]
```diff
@@ -2,6 +2,17 @@
 
 # coronet – Changelog
 
+## 4.2
+
+### Added
+- Incorporate custom event timestamps, i.e., add a configuration entry to the project configuration that allows specifying a file from which timestamps can be read, as well as an entry that allows locking this data; add corresponding functions `get.custom.event.timestamps`, `set.custom.event.timestamps` and `clear.custom.event.timestamps` (PR #227, 0aa342430ad3b354b9cf954dbe0838b056cf328a, 0f237d03913d2c940a008ea8fe84ba44817e77ea, c1803982357a3272b108f60cb1c976e3c2d9b1e5,
+54e089db0ceea07db94914d02655a7f1f67d3117, 54673f8f88ca276ba06396116d802425093544d4, c5f5403430d55ceff6b6d5acbbca1ae9c5c231e2)
+- Add function `split.data.time.based.by.timestamps` to allow using custom event timestamps for splitting. Alternatively, timestamps can be specified manually (PR #227, 051db9a5b81bf0d77805355611ef9f1a61f15b18, 5b8515f97da4a24f971be453589595d259ab1fa1, c5f5403430d55ceff6b6d5acbbca1ae9c5c231e2, 43f23a83bc66e485fea371f958bbb2ce3ddbd8d0)
```
Comment (bockthom) [inline NEWS.md:10]: I just checked the commit hashes mentioned here (prior to merging) and have two suggestions:
(1) Why do you mention commit 051db9a5b81bf0d77805355611ef9f1a61f15b18 at all? This behavior which is not present in any release and never was active as you have completely removed this afterwards. Therefore, I would suggest to just omit this commit hash in the news.
(2) Commit hash c5f5403430d55ceff6b6d5acbbca1ae9c5c231e2 is mentioned in both bullet points – this commit just changes the README. I understand why you have added the hash for "incorporate custom event timestamps", but I don't get why you mention this also in the bullet point regarding splitting – there is no change regarding splitting in the README, or do I miss something?

**Review by bockthom — APPROVED** [review]
Looks good. Approved now :smile:

**Pull request merged**